File: solutions/0144.Binary-Tree-Preorder-Traversal/Solution1.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
// recursion way.
let preorderTraversal = function (root) {
  let res = []
  helper(root, res)
  return res
}
function helper (root, res) {
  if (root === null) return
  res.push(root.val)
  helper(root.left, res)
  helper(root.right, res)
}

File: solutions/0144.Binary-Tree-Preorder-Traversal/Solution1.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// recursion way.
func preorderTraversal(root *TreeNode) []int {
    res := make([]int, 0)
    helper(root, &res)
    return res
}
func helper(root *TreeNode, res *[]int) {
    if root == nil {
        return
    }
    *res = append(*res, root.Val)
    helper(root.Left, res)
    helper(root.Right, res)
}
